zeroex/orderwatch/order_watcher_test.go

```diff
@@ -0,0 +1,666 @@
+	ethereumRPCMaxContentLength = 524288
+)
+
+var makerAddress = constants.GanacheAccount1
+var takerAddress = constants.GanacheAccount2
+var eighteenDecimalsInBaseUnits = new(big.Int).Exp(big.NewInt(10), big.NewInt(18), nil)
+var wethAmount = new(big.Int).Mul(big.NewInt(50), eighteenDecimalsInBaseUnits)
+var zrxAmount = new(big.Int).Mul(big.NewInt(100), eighteenDecimalsInBaseUnits)
+var tokenID = big.NewInt(1)
+
+// Since these tests must be run sequentially, we don't want them to run as part of
+// the normal testing process. They will only be run if the "--serial" flag is used.
+var serialTestsEnabled bool
+
+func init() {
+	flag.BoolVar(&serialTestsEnabled, "serial", false, "enable serial tests")
+	flag.Parse()
+}
+
+var rpcClient *ethrpc.Client
+var ethClient *ethclient.Client
+var zrx *wrappers.ZRXToken
+var dummyERC721Token *wrappers.DummyERC721Token
+var exchange *wrappers.Exchange
+var weth *wrappers.WETH9
+
+func init() {
+	var err error
+	rpcClient, err = ethrpc.Dial(constants.GanacheEndpoint)
+	if err != nil {
+		panic(err)
+	}
+	blockchainLifecycle, err = ethereum.NewBlockchainLifecycle(rpcClient)
+	if err != nil {
+		panic(err)
+	}
+	ethClient = ethclient.NewClient(rpcClient)
+	ganacheAddresses := ethereum.NetworkIDToContractAddresses[constants.TestNetworkID]
+	zrx, err = wrappers.NewZRXToken(ganacheAddresses.ZRXToken, ethClient)
+	if err != nil {
+		panic(err)
+	}
+	dummyERC721Token, err = wrappers.NewDummyERC721Token(constants.GanacheDummyERC721TokenAddress, ethClient)
+	if err != nil {
+		panic(err)
+	}
+	exchange, err = wrappers.NewExchange(ganacheAddresses.Exchange, ethClient)
+	if err != nil {
+		panic(err)
+	}
+	weth, err = wrappers.NewWETH9(ganacheAddresses.WETH9, ethClient)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func TestOrderWatcherUnfundedInsufficientERC20Balance(t *testing.T) {
+	if !serialTestsEnabled {
+		t.Skip("Serial tests (tests which cannot run in parallel) are disabled. You can enable them with the --serial flag")
+	}
+
+	teardownSubTest := setupSubTest(t)
+	defer teardownSubTest(t)
+
+	meshDB, err := meshdb.New("/tmp/leveldb_testing/" + uuid.New().String())
+	require.NoError(t, err)
+
+	signedOrder := scenario.CreateZRXForWETHSignedTestOrder(t, ethClient, makerAddress, takerAddress, wethAmount, zrxAmount)
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+	orderEventChan := setupOrderWatcherScenario(ctx, t, ethClient, meshDB, signedOrder)
+
+	// Transfer makerAsset out of maker address
+	opts := &bind.TransactOpts{
+		From:   makerAddress,
+		Signer: scenario.GetTestSignerFn(makerAddress),
+	}
+	txn, err := zrx.Transfer(opts, constants.GanacheAccount4, zrxAmount)
+	require.NoError(t, err)
+	ctx, cancelFn = context.WithTimeout(context.Background(), 4*time.Second)
+	receipt, err := bind.WaitMined(ctx, ethClient, txn)
+	cancelFn()
+	require.NoError(t, err)
+	assert.Equal(t, receipt.Status, uint64(1))
+
+	var orderEvents []*zeroex.OrderEvent
+	select {
```

Comment:
After seeing how many times this is used in the tests, maybe it's worth refactoring into a function? Something like:

```go
func waitForOrderEvents(t *testing.T, orderEventsChan chan<- []*zeroex.OrderEvent, timeout time.Duration) []*zeroex.OrderEvent {
	select {
	case orderEvents = <-orderEventsChan:
		return orderEvents
	case <-time.After(timeout):
		t.Fatal("timed out waiting for order events")
	}
} 
```